stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -616,166 +335,73 @@ private void releaseChildNodeRemovedEvent(ChildNodeRemovedEvent childNodeRemoved
+            element);
+      } else {
+        InspectNodeRequestedEvent message = new InspectNodeRequestedEvent();
+        message.nodeId = nodeId;
+        mPeerManager.sendNotificationToPeers("DOM.inspectNodeRequested", message);
+      }
     }
 
-    public void removeWithEvent(int index) {
-      Object element = remove(index);
+    public void onChildNodeRemoved(Object element) {
       ChildNodeRemovedEvent removedEvent = acquireChildNodeRemovedEvent();
-      removedEvent.parentNodeId = mParentNodeId;
-      removedEvent.nodeId = mObjectIdMapper.getIdForObject(element);
+
+      ElementInfo oldElementInfo = mDocument.getShadowDOM().getElementInfo(element);
+      removedEvent.parentNodeId = mDocument.getObjectIdMapper().getIdForObject(oldElementInfo.parentElement);
+      removedEvent.nodeId = mDocument.getObjectIdMapper().getIdForObject(element);
       mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", removedEvent);
+
       releaseChildNodeRemovedEvent(removedEvent);
     }
-  }
 
-  private final class DOMObjectIdMapper extends ObjectIdMapper {
-    @Override
-    protected void onMapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+    public void onChildNodeInserted(Object element, int parentNodeId, int previousNodeId, DOMView view, Accumulator<Object> insertedElements) {
+      ChildNodeInsertedEvent insertedEvent = acquireChildNodeInsertedEvent();
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.hook(object);
-    }
+      insertedEvent.parentNodeId = parentNodeId;
+      insertedEvent.previousNodeId = previousNodeId;
+      insertedEvent.node = createNodeForElement(view, element);
 
-    @Override
-    protected void onUnmapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+      insertedElements.store(element);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeInserted", insertedEvent);
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.unhook(object);
+      releaseChildNodeInsertedEvent(insertedEvent);
     }
   }
 
-  private final class ProviderListener implements DOMProvider.Listener {
-    @Override
-    public void onPossiblyChanged() {
-      updateTree();
-    }
-
+  private final class PeerManagerListener extends PeersRegisteredListener {
     @Override
-    public void onAttributeModified(Object element, String name, String value) {
-      mDOMProvider.verifyThreadAccess();
+    protected synchronized void onFirstPeerRegistered() {
+      // We don't call mDOMProvider.setListener() until the first DOM.getDocument request.
 
-      AttributeModifiedEvent message = new AttributeModifiedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      message.value = value;
-      mPeerManager.sendNotificationToPeers("DOM.attributeModified", message);
+      mDocument.createDomProviderIfNoneExists();
```

Comment:
This looks funky to me.  Why is the document not self managing?  Also, why is there not a document peer manager now?  It would seem that now that DOM and CSS are being intertwined together then the concept of them being enabled and the related peers also seem to be interconnected.  In other words, what would your code do if a caller invoked CSS.enable but not DOM.enable?